tcpdevmem.cu

```diff
@@ -0,0 +1,754 @@
+  for (cm = CMSG_FIRSTHDR(msg); cm; cm = CMSG_NXTHDR(msg, cm)) {
+    if (cm->cmsg_level != SOL_SOCKET ||
+        (cm->cmsg_type != SCM_DEVMEM_OFFSET &&
+          cm->cmsg_type != SCM_DEVMEM_HEADER)) {
+      continue;
+    }
+
+    devmemvec = (struct devmemvec *)CMSG_DATA(cm);
+
+    if (cm->cmsg_type == SCM_DEVMEM_HEADER) {
+      // TODO: process data copied from skb's linear
+      // buffer.
+      fprintf(stderr, "\n\nSCM_DEVMEM_HEADER. devmemvec->frag_size=%u\n",
+              devmemvec->frag_size);
+      exit(1);
+    }
+
+    struct devmemtoken token = { devmemvec->frag_token, 1 };
+    struct TcpdRxBlock blk;
+
+    blk.gpu_offset = (uint64_t)devmemvec->frag_offset;
+    blk.size = devmemvec->frag_size;
+    rx_blks_->emplace_back(blk);
+
+    total_received += devmemvec->frag_size;
+
+    vectors->emplace_back(*devmemvec);
+    tokens->push_back(token);
+  }
+
+  size_t dst_offset = tmbuf->bytes_received;
+  for (int i = 0; i < rx_blks_->size(); i++) {
+    struct TcpdRxBlock blk = rx_blks_->at(i);
+    size_t off = (size_t)blk.gpu_offset;
+    scattered_data_->emplace_back(
+        make_long3((long)dst_offset, (long)off, (long)blk.size));
+
+    dst_offset += blk.size;
+  }
+  tmbuf->bytes_received += received;
+
+  /* Once we've received fragments totaling buffer_size, we can copy from the
+   * CUDA buffer to a user-space buffer, and free the fragments in the CUDA
+   * buffer.
+  */
+  if (tmbuf->bytes_received == buffer_size) {
+    if (opts->tcpd_rx_cpy) {
+      gather_rx_data(tmbuf);
+      cudaDeviceSynchronize();
+    }
+    /* There is a performance impact when we cudaMemcpy from the CUDA buffer to
+     * the userspace buffer, so it's gated by a flag
+     */
+    if (opts->tcpd_validate) {
+      for (int idx = 0; idx < vectors->size(); idx++) {
+        struct devmemvec vec = (*vectors)[idx];
+        struct devmemtoken token = (*tokens)[idx];
+
+        /* copy each fragment to the cpy_buffer in order, i.e.
+         * 1st fragment will occuply bytes [0-4095], 2nd fragment will
+         * occupy bytes [4096-8191], etc.
+         */
+        cudaMemcpy(cpy_buffer + (vec.frag_token - 1) * PAGE_SIZE,
+                   (char *)tmbuf->gpu_gen_mem_ + vec.frag_offset,
+                   vec.frag_size,
+                   cudaMemcpyDeviceToHost);
+      }
+
+      /* Ensure the sequence is what we expect:
+       * a repeating sequence of 1 to LAST_PRIME inclusive
+       */
+      cudaDeviceSynchronize();
+      int i = 0;
+      int expected_val;
+      while (i < buffer_size) {
+        expected_val = (i % LAST_PRIME) + 1;
+        if (cpy_buffer[i] != expected_val) {
+          LOG_WARN(t->cb,
+                   "Thread %i - incorrect byte %i, expected %i, got %i",
+                   t->index,
+                   i,
+                   expected_val,
+                   cpy_buffer[i]);
+          break;
+        }
+        i++;
+      }
+    }
+
+    ret = setsockopt(client_fd, SOL_SOCKET,
```

Comment:
While not likely, there is a maximum size limit on the number of tokens we could pass to the kernel in one syscall. Consider using a loop and break the array into multiple calls if needed.